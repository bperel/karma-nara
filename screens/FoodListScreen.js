import React from 'react';
import Geo from '../lib/Geo';
import { ScrollView, StyleSheet, ActivityIndicator, FlatList, View, Text, Button, AsyncStorage } from 'react-native';
import Colors from '../constants/Colors';
import Coordinates from "../constants/Coordinates";

export default class FoodListScreen extends React.Component {
  static navigationOptions = {
    title: 'Karma nÃ¤ra',
  };

  state = {
    isLoadingComplete: false,
    locations: [],
  };

  fetchAndBuildLocations = () => (
    fetch(Coordinates.foodPlaceListUrl)
      .then((response) => response.json())
      .then((responseJson) => {
        let locations = responseJson
          .map(location => (
            {...location, distanceFromHQ: Geo.getDistanceFromKarmaHQ(location), following: false}
          ))
          .sort((location1, location2) => (
            location1.distanceFromHQ < location2.distanceFromHQ ? -1 : (location1.distanceFromHQ > location2.distanceFromHQ ? 1 : 0)
          ));
        this.setState({locations});
      })
      .catch((error) => {
        console.error(error);
      })
      .then(() => AsyncStorage.getItem('followedLocations'))
      .then(followedLocations => {
        let followedLocationIds = (JSON.parse(followedLocations) || []).map(followedLocation => followedLocation.id);
        this.setState(prevState => (
          {...prevState, locations: prevState.locations.map(location => (
            {...location, following: followedLocationIds.includes(location.id)}
          ))}
        ));
      })
      .finally(() => {
        this.setState({isLoadingComplete: true});
      })
  );

  componentDidMount() {
    return this.fetchAndBuildLocations()
  }

  toggleFollowFoodPlace(foodPlace) {
    this.setState(prevState => {
      prevState.locations.filter(location => location.id === foodPlace.id)[0].following = !foodPlace.following;
      AsyncStorage.setItem('followedLocations', JSON.stringify(this.state.locations.filter(location => !!location.following)));
      return prevState;
    });
  }

  isFollowing = locationId => !!this.state.locations.filter(location => location.id === locationId)[0].following;

  render() {
    return (
      <ScrollView style={styles.container}>
        {!this.state.isLoadingComplete && <ActivityIndicator size="large" color="#0000ff" />}
        {this.state.isLoadingComplete && <FlatList
          style={styles.flatlist}
          keyExtractor={foodPlace => foodPlace.id.toString()}
          data={this.state.locations}
          renderItem={({item: foodPlace}) =>
            <View style={styles.flatlistItem}>
              <View style={styles.follow}>
                <Button color={this.isFollowing(foodPlace.id) ? Colors.followButtonSelected : Colors.followButtonDefault}
                  title={ this.isFollowing(foodPlace.id) ? "Followed" : "Follow"}
                  onPress={() => { this.toggleFollowFoodPlace(foodPlace) }} />
              </View>
              <View style={styles.foodPlaceInfo}>
                <Text style={styles.name}>{foodPlace.name}</Text>
                <Text style={Geo.getProximityStyle(foodPlace.distanceFromHQ)}>{foodPlace.distanceFromHQ} meters away</Text>
              </View>
            </View>
          }
        />}
      </ScrollView>
    );
  }
}

const styles = StyleSheet.create({
  container: {
    flex: 1,
    backgroundColor: '#fff',
  },
  flatlist: {
    paddingTop: 15,
    paddingBottom: 15,
    paddingLeft: 25,
    paddingRight: 25,
    overflow: 'scroll',
  },
  flatlistItem: {
    paddingTop: 10,
    paddingBottom: 10,
    flexDirection: 'row',
    alignItems: 'center',
    borderRadius: 2,
  },
  follow: {
    width: 100,
  },
  foodPlaceInfo: {
    paddingLeft: 10,
    paddingRight: 10,
  },
  name: {
    fontSize: 18,
  },
});
