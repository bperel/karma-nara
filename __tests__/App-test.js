import 'react-native';
import React from 'react';
import { AsyncStorage as nativeStorage } from 'react-native'
import FoodListScreen from "../screens/FoodListScreen";

import Enzyme from 'enzyme';
import {shallow} from 'enzyme';
import Adapter from 'enzyme-adapter-react-16.3';

import fetchMock from 'fetch-mock';
import MockAsyncStorage from 'mock-async-storage';

Enzyme.configure({ adapter: new Adapter() });

fetchMock.get('*', JSON.stringify([
  {
    "id": 1,
    "name": "Karma canteen",
    "latitude": 59.330596,
    "longitude": 18.0560967,
    "following": false
  },
  {
    "id": 2,
    "name": "Teaterbaren",
    "latitude": 59.332147,
    "longitude": 18.065075,
    "following": true
  },
]));

jest.mock('AsyncStorage', () => new MockAsyncStorage());

describe('App snapshot', () => {
  it('resets locations to not followed by default', async () => {
    const wrapper = shallow(<FoodListScreen />);
    await wrapper.instance().componentDidMount();
    expect(wrapper.state('locations')[0].following).toMatchSnapshot();
    expect(wrapper.state('locations')[1].following).toMatchSnapshot();
  });
  it('calculates distances from HQ for each location', async () => {
    const wrapper = shallow(<FoodListScreen />);
    await wrapper.instance().componentDidMount();
    expect(wrapper.state('locations')[0].distanceFromHQ).toMatchSnapshot();
    expect(wrapper.state('locations')[1].distanceFromHQ).toMatchSnapshot();
  });

  it('handles already followed locations', async () => {
    await nativeStorage.setItem('followedLocations', JSON.stringify([{id: 2, following: true}]));
    const wrapper = shallow(<FoodListScreen />);
    await wrapper.instance().componentDidMount();
    expect(wrapper.state('locations')[1].following).toMatchSnapshot();
  });
});
