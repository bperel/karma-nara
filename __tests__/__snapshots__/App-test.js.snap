// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`App snapshot calculates distances from HQ for each location 1`] = `0`;

exports[`App snapshot calculates distances from HQ for each location 2`] = `539`;

exports[`App snapshot handles already followed locations 1`] = `true`;

exports[`App snapshot resets locations to not followed by default 1`] = `false`;

exports[`App snapshot resets locations to not followed by default 2`] = `false`;
